경북대 FE 장세현 2단계 - 로그인, 주문하기 API 구현하기

SeHIgh opened the pull request:
안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
2단계 과제 - '로그인, 주문하기 API 구현하기' 내용을 PR 요청 드립니다!

---

### 과제 진행 사항

#### 기존 요구사항

- 로그인 기능
  - `/login api` 를 사용하여 로그인 기능을 완성하기
  - 로그인 성공 시 내려오는 authToken과 email, name을 userInfo storage에 저장하고 활용하기
  - 4XX 에러가 발생하면 Toast를 통해 에러메시지를 보여주기. ( [react-toastify](https://www.npmjs.com/package/react-toastify) 라이브러리 사용)

- 주문하기 기능
  - `/products/:productId/summary` api를 사용하여 제품 정보를 가져오기
  - 만약 제품 정보 API에서 4XX 에러가 발생하면 Toast를 통해 에러메시지를 보여주고, 선물하기 홈으로 연결시키기
  - 보내는 사람 Input Field에 userInfo의 name을 defaultValue로 채우기
  - `/order` api를 사용하여 주문하기 기능을 완성하기
  - 주문하기 API의 경우 Authorization헤더에 로그인 응답에서 전달 받은 authToken을 넣어야만 동작하도록 구현하기
    - 이 부분은 인터셉터를 이용하여 구현하였습니다.
    (주문하기 전용 클라이언트가 해당 authToken 관련 인터셉터가 적용된 기존 apiClient를 그대로 상속 받도록 하였습니다.)
  - 주문하기 API에서 401 에러가 발생하면 로그인 페이지로 연결시키기

- 본인만의 기준으로 일관된 코드 작성
- 기능 단위로 나누어 커밋
- 재사용을 고려하여 API 구현

#### 추가 구현 사항

- 이전 1단계 에서 마지막에 주신 코멘트를 반영하여 초기에 보여줄 상품 개수의 상수 위치를 컴포넌트 위에 선언하도록 위치를 수정 반영하였습니다!
- 주문하기 성공 시 toast 알림 띄우도록 하였습니다.
- 인터셉터를 적극 활용하였습니다.
  - 주문하기 전용 클라이언트를 생성 후, 전용 인터셉터를 사용하여 401 에러 발생 시 로그인 페이지로 이동하도록 구현 하였습니다.
- 기존에 Context API 로 구현한 사용자 정보를 쿠키에 저장하고 해당 내용을 꺼내와서 쓰는 부분은 추후 합쳐서 개선해보고자 합니다!

### 질문 사항
- 현재 주문하기 API는 Authorization 헤더에 로그인 응답으로 전달받은 authToken을 반드시 포함해야 동작하도록 구현하는 것이 조건인데, 이를 위해 저는 apiClient 인스턴스의 인터셉터를 활용해 모든 요청에 authToken을 자동으로 포함하도록 설정했습니다.
저는 기본적으로 프론트엔드에서는 요청을 구분하지 않고, 서버에서 인증이 필요한 경우에만 토큰을 검증하는 구조를 생각하여 이처럼 구현하게 되었습니다.
그래서 만약 토큰이 제대로 전달이 안돼서 서버측에서 401 에러를 반환한다면, 로그인 페이지로 연결시키도록 별도 인스턴스에 인터셉터를 적용하였습니다.
이렇게 전역 인터셉터에 적용하는 방법이 괜찮은 접근 방식인지 아니면, 주문하기 API처럼 특정 인증이 필요한 요청만 별도의 Axios 인스턴스를 생성하고 해당 인스턴스에만 인터셉터를 적용하는 방식이 옳은 접근 방식인지 궁금합니다!

SeHIgh committed:
chore: 2차과제 요구사항 정리 및 react-toastify 라이브러리 추가

SeHIgh committed:
refactor(trend): 초기 보여줄 상품 개수 상수 위치 변경

SeHIgh committed:
feat(login-api): API를 이용한 로그인 기능 및 에러 처리 구현, 토스트 알림 추가

SeHIgh committed:
feat(order): 주문하기 페이지 API 적용

- 상품 요약 정보 조회 기능 추가

SeHIgh committed:
feat(order): 4XX 에러 발생 시 홈으로 자동 이동 기능 추가

SeHIgh committed:
feat(order): 주문하기 폼의 보내는 사람 Input Field에 userInfo의 name을 defaultValue로 설정

SeHIgh committed:
feat(order): 주문하기 기능 구현 및 API 요청 처리 추가

Leejin-Yang commented on src/pages/Order.tsx:
```diff
@@ -47,49 +57,46 @@ export const OrderContent = () => {
+  }
 
   // * 상품 정보가 없을 경우 NotFound 페이지로 이동하도록 처리
   if (!productInfo) return <NotFound />
```
데이터를 불러오는 중에 잠시 NotFound 페이지가 보이는데 로딩 상태를 사용해보면 어떨까요?

Leejin-Yang commented on src/api/orderClient.ts:
```diff
@@ -0,0 +1,29 @@
+
+// * 주문하기 전용 클라이언트 생성
+// ? 주문하기 전용 인터셉터를 적용하기위해!
+const orderClient = apiClient.create()
```
새로운 인스턴스를 생성할 때, 기존 apiClient에 설정한 인터셉트들이 상속되지 않아서 주문할 때 401 에러가 발생하고 있어요

Leejin-Yang replied:
질문에 대한 답변도 마저 하자면,

현재는 규모가 크지 않으니 apiClient만 사용해도 괜찮아보여요. 지금 구조에서는 401 에러가 발생할 때 리다이렉트도 같이 포함하면 더 좋아보이네요. 아니면 인증이 필요한 요청과 그렇지 않은 요청을 각각의 인스턴스로 만들어 관리해도 좋아보여요

규모가 커지고 다양한 곳에서 api를 불러야 하는 상황이나 각각의 다양한 처리가 필요하는 상황이 올 때, 책임을 분리해서 다중 인스턴스로 사용해 볼 수 있어보여요

Leejin-Yang replied:
인터셉트 기능도 적극적으로 활용하시고 많은 고민을 하신걸로 보여서 앞으로도 이렇게 학습해 나가시면 좋아보여요 👍

SeHIgh replied:
이전에 apiClient를 사용하였을때는 에러발생 없이 잘돼가지고 당연히 인터셉터도 상속이 된다 판단하고, 끝을 냈던 것 같습니다.
끝내기 전에 다시한번 검토하는 습관을 가지도록 하겠습니다!

---

제가 인스턴스를 분리하게된 계기는
현재 규모가 크진 않지만, ui 쪽에서 특정 에러(401 Error)에 대한 처리를 하려니 코드가 비교적 복잡해지는 문제가 있는 것 같아 
api 딴에서 인증 전용 인스턴스를 분리하고 해당 인스턴스에 인터셉터를 이용하여 관리하는 방향으로 생각을 하였습니다!

Leejin-Yang commented on src/pages/Login.tsx:
```diff
@@ -33,17 +34,19 @@ export const Login = () => {
+    })
 
     // * 로그인 정보 저장 (쿠키에 암호화되어 저장)
+    // TODO: 이 부분은 이전에 context api를 활용하여 안전하게 쿠키에 저장 시에 사용하던 로직으로
+    // TODO: 이번에는 남겨두고(일단 이번 미션에 집중) api 쪽 코드에 추후 적용할 예정
```
👍

Leejin-Yang commented on src/pages/Order.tsx:
```diff
@@ -47,49 +57,46 @@ export const OrderContent = () => {
     if (!productInfo) return
 
     if (data.receivers.length === 0) {
       alert('받는 사람이 최소 1명 필요합니다.')
```
이 부분도 토스트를 적용할 수 있겠네요 😊

Leejin-Yang commented on src/contexts/order/OrderFormContext.tsx:
```diff
@@ -20,13 +20,19 @@ type OrderFormContextType = {
 
 // * 주문 폼 Provider
-export function OrderFormProvider({ children }: { children: ReactNode }) {
+export function OrderFormProvider({
```
이건 이번 단계를 너무 잘 진행해주셔서.. 외적인 피드백인데요, 모달 상태는 컨텍스트를 분리해보면 어떨까요?

지금은 모달이 한개이니 괜찮지만, 앞으로 여러 모달이 생긴다면 재사용할 수 있도록 모달 컨텍스트를 만드는 시도를 해보는 것도 좋아보여요 😊

SeHIgh replied:
조언 감사합니다!
해당 형태를 컨텍스트로 분리해서 코드의 중복을 줄이고, 여러 모달의 상태를 효율적으로 관리할 수 있도록 개선 해보도록 하겠습니다!

Leejin-Yang reviewed (changes requested):
세현님 안녕하세요 👋 이번 단계도 잘 진행해주셨네요
잘 진행해주셔서 자잘한 코멘트를 몇개 남겨두었는데, 확인 부탁드려요 😊
그리고 주문 요청 관련 코멘트는 꼭 확인해주세요.!
수정해주시고 다시 요청 주세요

SeHIgh committed:
refactor(toastify): 경고 알림을 alert 에서 toast로 통일

- 받는 사람 관련 경고 메시지를 toast로 변경

SeHIgh committed:
feat(order): 주문 페이지 로딩 상태 처리

- 로딩 중 화면 추가
- 초기화 완료 전 아무것도 렌더링하지 않도록 수정
- 상품 정보가 없을 경우와 에러 발생을 동일하게 판단하여 중복된 코드 제거

SeHIgh committed:
refactor(api): authClient 인스턴스로 인증 필요 요청 관리

- apiClient 는 기본 인스턴스로 에러 처리 인터셉터만 남기고 인증 관련 인터셉터는 authClient로 이전
- 주문 API 요청을 authClient를 통해 처리하도록 수정
- 에러 처리 로직 수정 및 토스트 메시지 통일

SeHIgh commented on src/api/authClient.ts:
```diff
@@ -0,0 +1,46 @@
+// ? apiClient 인스턴스 기본 설정 상속
+const authClient = apiClient.create()
+
+// * 토큰 인터셉터
+authClient.interceptors.request.use(
+  (config) => {
+    const storedUser = localStorage.getItem(STORAGES.AUTH)
+    if (storedUser) {
+      const { authToken } = JSON.parse(storedUser)
+
+      if (authToken) {
+        config.headers.Authorization = `${authToken}`
+      }
+    }
+    return config
+  },
+  (error) => Promise.reject(error),
+)
+
+// * 에러 처리 인터셉터
+authClient.interceptors.response.use(
+  (response) => response,
+  async (error) => {
+    const status = error.response?.status
+    const message = error.response?.data?.data?.message
+
+    // ! 401 에러가 발생하면 로그인 페이지로 연결
+    if (status === 401) {
+      toast.error(message)
+      // * 로그인 페이지로 리다이렉트
+      window.location.href = ROUTE_PATH.LOGIN
+    } else if (status && status >= 400 && status < 500) {
+      toast.error(message)
+    }
+
+    return Promise.reject(error)
+  },
+)
+
+export default authClient
```
주문 전용 인스턴스에서 인증 전용 인스턴스로 명칭을 바꾸고 인증을 요구하는 요청은 해당 인스턴스로 전부 이전하고, 이를 사용하는 요청은 헤더에 인증 정보를 담아 요청하도록 하였습니다.

SeHIgh replied:
이전에 인스턴스를 상속하면, 그에 적용된 인터셉터도 같이 상속이 되는줄 알았는데,
다시 알아보니 인스턴스 정보는 상속이 되지만, 인터셉터는 상속이 되지 않는다는 것을 알게 되어
이처럼 아예 해당 기능을 이전하여 관리하도록 수정하였습니다.

SeHIgh replied:
따라서 이전에 apiClient에 있던 인증 관련 인터셉터는 제거하였습니다

SeHIgh commented on src/pages/Order.tsx:
```diff
@@ -47,52 +57,55 @@ export const OrderContent = () => {
+    navigate(ROUTE_PATH.HOME)
+    return null
+  }
 
```
정보가 없을 경우와 에러 발생 조건이 같다 판단하여 해당 내용을 없애고 에러 발생 시 처리 부분만 남겼습니다.

SeHIgh commented on src/pages/Order.tsx:
```diff
@@ -47,52 +57,55 @@ export const OrderContent = () => {
-  const totalPrice = productInfo ? getTotalPrice(productInfo.price.sellingPrice) : 0
+  const totalPrice = productInfo ? getTotalPrice(productInfo.price) : 0
+
+  // * 로딩 중 화면
+  if (isLoading) {
+    return (
+      <LoadingContainer>
+        <Loading />
+      </LoadingContainer>
+    )
+  }
```
로딩 처리가 없어 이전의 NotFound 페이지가 표시가 되었던 것 같습니다. 로딩 중일 때 렌더링할 요소를 추가하여 해당 부분을 해결하였습니다!

SeHIgh reviewed (commented):
멘토님 께서 남겨주셨던 코멘트들을 반영하여 수정해보았습니다!

1.  주문하기 페이지 (Order.tsx) 데이터 불러오는 중에 잠시 NotFound 페이지가 보이는 현상 해결하기  
	-> isLoading 를 이용하여 해결하였습니다!

2. 새로운 인스턴스 생성 시, 기존 apiClient에 설정한 인터셉트들이 상속되지 않는 문제 발생(주문 시 401 에러) 해결하기
	-> 기본적으로 인스턴스 정보는 상속이 되지만, 인터셉터 정보는 상속이 되지 않는 다는 것을 알게 되었습니다. (이전에는 인터셉터도 같이 상속되는 줄로 잘못 알 고 있었습니다...)
	=> 인증 전용 인스턴스 와 기본 인스턴스 두개로 나누어 관리하고, 인증 관련한 내용을 완전히 이전 시켰습니다!

3. onSubmit에 alert로 된 부분 toast로 통일하기

Leejin-Yang commented on src/pages/Order.tsx:
```diff
@@ -62,7 +62,7 @@ export const OrderContent = () => {
 
     if (data.receivers.length === 0) {
-      alert('받는 사람이 최소 1명 필요합니다.')
+      toast.warning('받는 사람이 최소 1명 필요합니다.')
```
👍

Leejin-Yang commented on src/api/index.ts:
```diff
@@ -11,32 +10,15 @@ const apiClient = axios.create({
   async (error) => {
     const status = error.response?.status
-    const message = error.response?.data?.data?.message || '요청 처리 중 오류가 발생했습니다.'
+    const message = error.response?.data?.data?.message
```
디폴트 값을 지우셨네요 👀
메시지가 꼭 담겨올지 모르니 있어도 좋아보여요

Leejin-Yang commented on src/api/authClient.ts:
```diff
@@ -0,0 +1,46 @@
+
+// * 인증이 필요한 요청용 클라이언트 생성
+// ? apiClient 인스턴스 기본 설정 상속
+const authClient = apiClient.create()
```
잘 분리해주셨네요 👍

Leejin-Yang reviewed (approved):
빠르게 반영해주셨네요 👍
전체적으로 구조도 잘 잡아주셨고, 요구사항도 다 만족해주셔서 이번 단계도 잘 진행해주신 것으로 보여요 😊
여기서 마무리하고 다음 단계에서 보아요 🙌
